bee-protocol/src/worker/confirmer.rs

```diff
@@ -0,0 +1,202 @@
+where
+    N::Backend: StorageBackend,
+{
+    type Config = ();
+    type Error = Infallible;
+
+    fn dependencies() -> &'static [TypeId] {
+        vec![TypeId::of::<TangleWorker>()].leak()
+    }
+
+    async fn start(node: &mut N, _config: Self::Config) -> Result<Self, Self::Error> {
+        let (tx, rx) = mpsc::unbounded_channel();
+
+        let tangle = node.resource::<MsTangle<N::Backend>>();
+
+        node.spawn::<Self, _, _>(|shutdown| async move {
+            info!("Running.");
+
+            let mut receiver = ShutdownStream::new(shutdown, UnboundedReceiverStream::new(rx));
+
+            while let Some(ConfirmationWorkerEvent(index, milestone)) = receiver.next().await {
+                process(&tangle, milestone, index).await;
+            }
+
+            // Before the worker completely stops, the receiver needs to be drained for milestone cones to be updated.
+            // Otherwise, information would be lost and not easily recoverable.
+
+            let (_, mut receiver) = receiver.split();
+            let mut count: usize = 0;
+
+            while let Some(Some(ConfirmationWorkerEvent(index, milestone))) = receiver.next().now_or_never() {
+                process(&tangle, milestone, index).await;
+                count += 1;
+            }
+
+            debug!("Drained {} milestones.", count);
+
+            info!("Stopped.");
+        });
+
+        Ok(Self { tx })
+    }
+}
+
+async fn process<B: StorageBackend>(tangle: &MsTangle<B>, milestone: Milestone, index: MilestoneIndex) {
+    let message_id = milestone.message_id();
+
+    if let Some((parent1, parent2)) = tangle
+        .get(message_id)
+        .await
+        .map(|message| (*message.parent1(), *message.parent2()))
+    {
+        // Confirm the past cone of this milestone, and return all newly confir
+        let confirmed = confirm_past_cone(tangle, parent1, parent2, index).await;
+
+        // Propagate new confirmation states
+        update_otrsi_ytrsi(tangle, confirmed).await;
+    }
+
+    // Update tip pool after all values got updated.
+    tangle.update_tip_scores().await;
+}
+
+async fn confirm_past_cone<B: StorageBackend>(
+    tangle: &MsTangle<B>,
+    parent1: MessageId,
+    parent2: MessageId,
+    index: MilestoneIndex,
+) -> HashSet<MessageId> {
+    let mut parents = vec![parent1, parent2];
+    let mut confirmed = HashSet::new();
+
+    while let Some(parent_id) = parents.pop() {
+        // Our stop conditions. Note that the order of calls is important (from cheap to more expensive) for performance
+        // reasons.
+        if confirmed.contains(&parent_id)
+            || tangle.is_solid_entry_point(&parent_id)
+            || tangle.get_metadata(&parent_id).await.unwrap().flags().is_confirmed()
+        {
+            continue;
+        }
+
+        tangle
+            .update_metadata(&parent_id, |metadata| {
+                // TODO: Throw one of those indexes away ;)
+                metadata.set_milestone_index(index);
```

Comment:
in the old "milestone_cone_updater" we didn't set the `milestone_index`.
this is something that is done by bee-ledger since it also traverses the past cone (duplicate work)
this might conflict with bee-ledger traversal if we have it in there (so maybe we should remove this).

Reply:
nope, the ledger is not setting it